File: 0042/solution.ts
function trap(height: number[]): number {
  let H = height.length;
  let res = 0;
  let leftMax = 0;
  const rightMax: Array<number> = new Array(H).fill(0);
  for (let i = H - 2; i >= 0; i--) {
    rightMax[i] = Math.max(rightMax[i + 1], height[i + 1]);
  }

  for (let i = 1; i < H; i++) {
    leftMax = Math.max(leftMax, height[i - 1]);
    const contribution = Math.min(leftMax, rightMax[i]);
    res += Math.max(0, contribution - height[i]);
  }
  return res;
}

File: 0042/solution.py
class Solution:
    def trap(self, height: list[int]) -> int:
        trapped = 0
        left, right = 0, len(height) - 1
        left_max, right_max = height[left], height[right]

        while left < right:
            if height[left] < height[right]:
                if height[left] >= left_max:
                    left_max = height[left]
                else:
                    trapped += left_max - height[left]
                left += 1
            else:
                if height[right] >= right_max:
                    right_max = height[right]
                else:
                    trapped += right_max - height[right]
                right -= 1

        return trapped
